### PR TITLE
Added batching rule for sdpa_math, sdpa_efficient_attention forward, cudnn, and flash attention

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133964

Fixes https://github.com/pytorch/pytorch/issues/117016, https://github.com/pytorch/pytorch/issues/102457, https://github.com/pytorch/pytorch/issues/110525, https://github.com/pytorch/pytorch/issues/108065, 

cc @zou3519

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesConvolution.cpp [hunk 1]
```diff
@@ -138,81 +138,6 @@ static Tensor _convolution_decomp(
       input_r, weight_r, bias_r_opt, stride_, padding_, dilation_, transposed_, output_padding_, groups_);
 }
 
-// TODO: delete the following after confirming performance
```
Comment (Chillee) [inline aten/src/ATen/functorch/BatchRulesConvolution.cpp:141]: I presume we've confirmed performance by now :P

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 2]
```diff
@@ -485,6 +485,49 @@ pinv_batch_rule(
...
+  if (attn_bias.has_value() && attn_bias->defined()) {
+    attn_bias_ = attn_bias_bdim.has_value() ? reshape_dim_into(*attn_bias_bdim, 0, attn_bias.value()) : attn_bias.value();
+  }
+  const auto [res0, res1, res2, res3] = at::_scaled_dot_product_efficient_attention(
+      query_, key_, value_, attn_bias_, compute_log_sumexp, dropout_p, is_causal, scale);
+  const auto batch_size = extra_batch_size;
+  const auto res0_ = reshape_dim_outof(0, batch_size, res0);
+  const auto res1_ = reshape_dim_outof(0, batch_size, res1);
+  // philox seed is always non-batched
+  return std::make_tuple(res0_, 0, res1_, 0, res2, std::nullopt, res3, std::nullopt);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:528]: The tensor args in the tuple can have std::move called on them to avoid a copy and therefore atomic lock.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Plan on doing another PR for flash attention and cudnn attention?

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 3]
```diff
@@ -485,6 +485,49 @@ pinv_batch_rule(
     const std::optional<int64_t> rtol_bdim, bool hermitian) {
   return atol_rtol_tensor_batch_rule(ATEN_FN2(linalg_pinv, atol_rtol_tensor), input, input_bdim, atol, atol_bdim, rtol, rtol_bdim, hermitian, "linalg.pinv");
 }
+
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:488]: Can you add tests? It's suspicious that none of the xfails in test_vmap flipped

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 4]
```diff
@@ -485,6 +485,49 @@ pinv_batch_rule(
...
+  std::optional<Tensor> attn_bias_;
+  if (attn_bias.has_value() && attn_bias->defined()) {
+    attn_bias_ = attn_bias_bdim.has_value() ? reshape_dim_into(*attn_bias_bdim, 0, attn_bias.value()) : attn_bias.value();
+  }
+  const auto [res0, res1, res2, res3] = at::_scaled_dot_product_efficient_attention(
+      query_, key_, value_, attn_bias_, compute_log_sumexp, dropout_p, is_causal, scale);
+  const auto batch_size = extra_batch_size;
+  const auto res0_ = reshape_dim_outof(0, batch_size, res0);
+  const auto res1_ = reshape_dim_outof(0, batch_size, res1);
+  // philox seed is always non-batched
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:527]: Is there randomness involved in sdpa? Or is it deterministic and just takes a seed?

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 3 above)
Comment (Chillee) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:488]: Yeah the SDPA tests are skipped because of randomness.

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 5]
```diff
@@ -485,6 +485,49 @@ pinv_batch_rule(
...
+  auto key_ = moveBatchDimToFront(key, key_bdim);
+  auto value_ = moveBatchDimToFront(value, value_bdim);
+  int64_t extra_batch_size = 1;
+  if (query_bdim.has_value()) {
+    extra_batch_size = query.size(0);
+  } else if (key_bdim.has_value()) {
+    extra_batch_size = key.size(0);
+  } else if (value_bdim.has_value()) {
+    extra_batch_size = value.size(0);
+  }
```
Comment (guilhermeleobas) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:509]: Shouldn't here be `query_`, `key_` and `value_`?

```suggestion
  if (query_bdim.has_value()) {
    extra_batch_size = query_.size(0);
  } else if (key_bdim.has_value()) {
    extra_batch_size = key_.size(0);
  } else if (value_bdim.has_value()) {
    extra_batch_size = value_.size(0);
  }
```

Otherwise, code like this will fail:

```python
import torch
import torch.nn.functional as F

def fn(query, key, value):
    return F.scaled_dot_product_attention(query, key, value)

B = 4
query = torch.rand(32, B, 8, 128, 64, dtype=torch.float16, device="cuda")
key = torch.rand(B, 32, 8, 128, 64, dtype=torch.float16, device="cuda")
value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
# y = F.scaled_dot_product_attention(query,key,value)
y = torch.vmap(fn, in_dims=(1, 0, None))(query, key, value)
print(y.shape)
```

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
If the input tensors have less than 4 dimensions + batched dim, PyTorch will still raise a performance warning:

```python
import torch
import torch.nn.functional as F

def fn(query, key, value):
    return F.scaled_dot_product_attention(query, key, value)

B = 4
query = torch.rand(32, B, 8, 128, dtype=torch.float16, device="cuda")
key = torch.rand(B, 32, 8, 128, dtype=torch.float16, device="cuda")
value = torch.rand(32, 8, 128, dtype=torch.float16, device="cuda")
y = torch.vmap(fn, in_dims=(1, 0, None))(query, key, value)
print(y.shape)


# /home/guilhermeleobas/git/pytorch/a.py:88: UserWarning: There is a performance
# drop because we have not yet implemented the batching rule for aten::_safe_softmax.
# Please file us an issue on GitHub so that we can prioritize its implementation.
# (Triggered internally at /home/guilhermeleobas/git/pytorch/aten/src/ATen/functorch/BatchedFallback.cpp:81.)
#   return F.scaled_dot_product_attention(query, key, value)
```

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 5 above)
Comment (Chillee) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:509]: Good catch!

**Comment by Chillee** [comment]
@guilhermeleobas yeah, we recently added a `_safe_softmax` path to the MATH path - I'll add a batching rule for softmax too.

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:564]: There is randomness, we return the seeds from forwards to backwards though

File: aten/src/ATen/functorch/BatchRulesConvolution.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesConvolution.cpp:141]: lol yes

File: test/functorch/test_vmap.py [hunk 6]
```diff
@@ -3850,6 +3850,64 @@ def test_threshold(self, device):
...
 
+    def test_sdpa(self, device):
+        if device == "cpu":
+            raise unittest.SkipTest("This test is only for CUDA for now")
+        from torch.nn.attention import sdpa_kernel, SDPBackend
+
+        backends = [
+            # SDPBackend.MATH,
+            # SDPBackend.EFFICIENT_ATTENTION,
+            SDPBackend.FLASH_ATTENTION  # not supported yet
```
Comment (zou3519) [inline test/functorch/test_vmap.py:3861]: are the ones that are commented out not supported? or is FLASH_ATTENTION not supported?

File: test/functorch/test_vmap.py [hunk 7]
```diff
@@ -3850,6 +3850,64 @@ def test_threshold(self, device):
...
+            backend_ctx = sdpa_kernel([backend])
+            with backend_ctx:
+                for batching in [
+                    (True, True, True),
+                    (True, False, False),
+                    (False, True, True),
+                ]:
+                    size = [8, 4, 128, 64]
+                    if batching[0]:
+                        query = T(3, *size)
```
Comment (zou3519) [inline test/functorch/test_vmap.py:3878]: nit: B = 3 for clarity

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:563]: @Chillee does vmap over _scaled_dot_product_efficient_attention with "randomness=None" error out? In general we expect random operations to error out with "randomness=None" and to do something reasonable with the other randomness schemes

Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:563]: discussed offline: me to figure out how to actually say "when p=0 then we don't need a randomness option".

Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:563]: The way to do this is, inside the batching rule (no need to change the plumbing): get `maybeCurrentDynamicLayer()->randomness()` and then handle it. That is, if p > 0, check that randomness is "diferent" (if not, error out). Otherwise, if p = 0, then do nothing.

File: test/functorch/test_vmap.py (hunk 6 above)
Comment (Chillee) [reply test/functorch/test_vmap.py]: oops

**Comment by drisspg** [comment]
@Pytorchbot merge
